apollo-router/src/configuration/cors.rs

```diff
@@ -84,90 +154,338 @@ impl Cors {
+                true,
+                request_headers,
+            );
+            return Box::pin(async move { Ok(response) });
+        }
+
+        let fut = self.inner.call(req);
+        Box::pin(async move {
+            let mut response = fut.await?;
+            // Find matching origin configuration
+            let origin_config = Self::find_matching_origin_config(&config, &origin);
+            // Add CORS headers for non-preflight
+            Self::add_cors_headers(
+                &mut response,
+                &config,
+                &origin_config,
+                &origin,
+                false,
+                request_headers,
+            );
+            Ok(response)
+        })
+    }
+}
+
+impl<S> CorsService<S> {
+    /// Find the matching origin configuration for a given origin
+    fn find_matching_origin_config<'a>(
+        config: &'a Cors,
+        origin: &'a Option<http::HeaderValue>,
+    ) -> Option<&'a OriginConfig> {
+        let origin_str = origin.as_ref()?.to_str().ok()?;
+
+        // Check for exact origin matches first
+        for origin_config in &config.origins {
+            for url in &origin_config.origins {
+                if url == origin_str {
+                    return Some(origin_config);
+                }
+            }
+
+            // Check regex matches
+            if !origin_config.match_origins.is_empty() {
+                for regex in &origin_config.match_origins {
+                    if regex.is_match(origin_str) {
+                        return Some(origin_config);
+                    }
+                }
+            }
+        }
+
+        None
+    }
+
+    /// Add CORS headers to the response
+    fn add_cors_headers<ResBody>(
+        response: &mut Response<ResBody>,
+        config: &Cors,
+        origin_config: &Option<&OriginConfig>,
+        origin: &Option<http::HeaderValue>,
+        is_preflight: bool,
+        request_headers: Option<http::HeaderValue>,
+    ) {
+        // Determine which configuration to use (origin-specific or global)
+        let allow_credentials = origin_config
+            .and_then(|oc| oc.allow_credentials)
+            .unwrap_or(config.allow_credentials);
+
+        let allow_headers = origin_config
+            .and_then(|oc| {
+                if oc.allow_headers.is_empty() {
+                    None
+                } else {
+                    Some(&oc.allow_headers)
+                }
+            })
+            .unwrap_or(&config.allow_headers);
+
+        let expose_headers = origin_config
+            .map(|oc| oc.expose_headers.as_ref())
+            .or(config.expose_headers.as_ref());
+
+        let methods = origin_config
+            .and_then(|oc| {
+                if oc.methods.is_empty() {
+                    None
+                } else {
+                    Some(&oc.methods)
+                }
+            })
+            .unwrap_or(&config.methods);
+
+        let max_age = config.max_age;
```

Comment:
We dont have a neeed for this, but could other customers have a use case to have max_age be potentially be able to be set differently in each originconfig?